docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,324 @@
+      ups.head
+    }
+  )
+  lazy val module = new LazyModuleImp(this) {
+    require(node.in.size == 2)
+    node.out.head._1 := node.in.head._1 + node.in.tail.head._1
+  }
+
+  override lazy val desiredName = "Adder"
+}
+```
+
+Our `AdderDriver` randomly generates an addend of `width` bits each cycle, and
+passes it to `numOutputs` sources.
+
+```scala mdoc
+/** driver (source)
+  * drives one random number on multiple outputs */
+class AdderDriver(width: Int, numOutputs: Int = 2)(implicit p: Parameters) extends LazyModule {
+  val nodeSeq = Seq.fill(numOutputs)(new AdderDriverNode(DownwardParam(width)))
+
+  lazy val module = new LazyModuleImp(this) {
+    // generate random addend (notice the use of the negotiated width)
+    val randomAddend = FibonacciLFSR.maxPeriod(nodeSeq.head.out.head._2.width)
+
+    // drive signals
+    nodeSeq.foreach { node => node.out.head._1 := randomAddend }
+  }
+
+  override lazy val desiredName = "AdderDriver"
+}
+```
+
+Our `AdderMonitor` prints out the results of the `Adder` each cycle and
+signals an error if the `Adder` returns an incorrect result. It has two
+`AdderMonitorNode`s for taking in addends from the `AdderDriver`, and one
+`AdderMonitorNode` to take in the sum from the `Adder`.
+
+```scala mdoc
+/** monitor (sink) */
+class AdderMonitor(width: Int)(implicit p: Parameters) extends LazyModule {
+  val nodeSeq = Seq.fill(2) { new AdderMonitorNode(UpwardParam(width)) }
+  val nodeSum = new AdderMonitorNode(UpwardParam(width))
+
+  lazy val module = new LazyModuleImp(this) {
+    val io = IO(new Bundle {
+      val error = Output(Bool())
+    })
+
+    // print output
+    printf(p"${nodeSeq(0).in.head._1} + ${nodeSeq(1).in.head._1} = ${nodeSum.in.head._1}")
+
+    // basic correctness checking
+    io.error := nodeSum.in.head._1 =/= nodeSeq(0).in.head._1 + nodeSeq(1).in.head._1
+  }
+
+  override lazy val desiredName = "AdderMonitor"
+}
+```
+
+### Creating the Top
+
+Our top-level module will be a test bench that instantiate our `Adder` and its
+peripherals. Notice that different `width` parameters are passed into Diplomacy
+via `drivers` and `checker`. As described in our `AdderNodeImp`, we expect the
+smaller of the two widths to actually be used.
+
+In this module, we also bind our nodes. Sinks are on the left-hand side, while
+sources are on the right.
+
+```scala mdoc
+/** top-level connector */
+class AdderTestHarness()(implicit p: Parameters) extends LazyModule {
+  val adder = LazyModule(new Adder)
+  val drivers = Seq.fill(2) { LazyModule(new AdderDriver(width = 8)) } // 8 will be a downward-traveling width
+  val checker = LazyModule(new AdderMonitor(width = 4))                // 4 will be an upward-traveling width
+
+  // create edges via binding operators between nodes in order to define a complete graph
+  drivers.foreach{ driver => adder.node := driver.nodeSeq(0) }
+
+  drivers.zip(checker.nodeSeq).foreach { case (driver, checkerNode) => checkerNode := driver.nodeSeq(1) }
```

Comment:
`driver.nodeSeq(0)` and `driver.nodeSeq(1)` both become `driver.node` if you take my above suggestion

Reply:
🙏 that looks SO much nicer